keps/sig-cli/kubectl-apply-prune-enhancements.md

```diff
@@ -0,0 +1,273 @@
+ 
+- Pruning is widely used in CI workflow with Kubernetes.
+
+## Implementation History
+
+The major implementation includes following steps:
+
+- add functions to identify and validate apply-config object
+- add a different code path to handle the case when apply-config object is present
+- add unit test
+- add e2e tests
+- update `kubectl apply` document and examples
+
+As a convenient way for users to view the intention of apply, `--dry-run` will show categorized output: 
+```
+$ kubectl apply -f <dir> --dry-run
+Resources to add
+  pod1...
+  pod2...
+Resources modified
+  pod3...
+  pod4...
+Resources unmodified
+  pod5...
+  pod6...
+Resources to delete
+  pod7...     
+```
+
+## Drawbacks
+
+- This approach is not backward compilable with Alpha `--prune`
+- The apply-config object can't be auto pruned. It need manual clean up.
+
+## Alternatives
+
+### Using a unique identifier as label
+Instead of listing all the child objects in the apply-config object, a label could be set in the apply-config object. 
+```
+apiVersion: v1
+kind: ConfigMap
+metadata:
+ name: predictable-name-foo
+ labels:
+   kubectl.kubernetes.io/apply-confg: "true"
+   kubectl.kubernetes.io/apply-id: some_uuid
+```
+
+All child objects have this label. Then `kubectl apply --prune` is run, the live set of child objects is found by label selecting and annotation matching.
+```
+ labels:
+   kubectl.kubernetes.io/apply-id: some_uuid
+ annotations:
+    kubectl.kubernetes.io/apply-owner: uuid_of_the_apply_config_object
+```
+Any object in the live set that is not in the new apply set will be deleted.
+
+The label set in the apply-config object should not have collisions with resources outside of the aggregation identified by the apply-config objects. This will require users or tools create a unique identifier.
+- pros:
+  - This extends the idea of Alpha `--prune` with an apply-config object; existing `--prune` users can easily adopt this new approach.
+  - This could be used by other functionality of kubectl, such as checking if the applied objects are settled
+  - This can work for Application CRD or other CRDs that with an apply-config object concept
+  - No need to maintain a list of child objects and store it in the apply-config object
+- cons:
+  - need a unique label that doesn't collision with other resources; may need other tool to mechanism to generate a unique label
+  - selecting matching objects across namespaces and types is not efficient
+  - The unique identifier may not be human-readable
+
+### Handling prune on the server side
+Currently, some logic in `kubectl apply` has started to be moved to the server side. `kubectl apply` performs 3-way merge for resolving state differences before sending the state to API server. With alpha release of [server-side `apply`](https://github.com/kubernetes/enhancements/issues/555), this 3-way merge is going to be replaced. Similar to that, the pruning logic in `kubectl apply` can also be moved to the server side.
```

Comment:
Cross-object apply was explicitly out of scope. apiserver doesn't currently do any cross-object operations. So, I don't actually see how this can be done (without some radical apiserver scope-creep).

Additionally, the hard problem is tracking which apply invocation is the successor to which other apply invocation; moving the feature to the control plane doesn't address that problem at all, you still need some solution.

Reply:
Another alternative is to still host the pruning logic in server side, but outside of `kubectl apply`.